strest-client crashes with a goroutine panic: `index out of range`

```
stevej@proxy-test-4b:~$ ./strest-client -address 10.240.0.4:4140 -latencyPercentiles 100=10 -lengthPercentiles 100=10 -requests 10000 -concurrency 200
2016-08-08T20:59:10Z  11669/0/0 requests 10s   2 [153 344 531  983 ] 1468
2016-08-08T20:59:21Z  29065/0/0 requests 10s   1 [ 63 155 212  431 ]  431
2016-08-08T20:59:31Z  38276/0/0 requests 10s   0 [ 50 103 130  228 ]  228
panic: runtime error: index out of range

goroutine 47 [running]:
panic(0x88f560, 0xc82000e080)
        /usr/local/go/src/runtime/panic.go:481 +0x3e6
math/rand.(*rngSource).Int63(0xc82016a000, 0x7b0480)
        /usr/local/go/src/math/rand/rng.go:243 +0xc8
math/rand.(*Rand).Int63(0xc8201292b0, 0xc824674d01)
        /usr/local/go/src/math/rand/rand.go:46 +0x39
math/rand.(*Rand).Int(0xc8201292b0, 0xc82014c360)
        /usr/local/go/src/math/rand/rand.go:56 +0x21
main.main.func2(0xc820129140, 0xc820129190, 0xc82012e228, 0xc8201292b0, 0xc82012e220, 0xc82014c360, 0xc820128010)
        /Users/stevej/src/strest-grpc/client/main.go:149 +0x2bd
created by main.main
        /Users/stevej/src/strest-grpc/client/main.go:165 +0xc90
```

--- a/client/distribution/distribution.go
+++ b/client/distribution/distribution.go
@@ -7,24 +7,20 @@ import (
 	"sort"
 )
 
-/**
- * Distribution is a map of percentiles (multiplied by 10) to values.
- * e.g. a latency distribution might be
- * 500  -> 1
- * 900  -> 20
- * 950  -> 30
- * 990  -> 50
- * 999  -> 100
- * 1000 -> 100
- *
- * When a value is requested that no exact value is known for, we'll
- * calculate it's value as a linear extrapolation.
- */
+// Distribution is a map of percentiles (multiplied by 10) to values.
+// e.g. a latency distribution might be
+// 500  -> 1
+// 900  -> 20
+// 950  -> 30
+// 990  -> 50
+// 999  -> 100
+// 1000 -> 100
+//
+// When a value is requested that no exact value is known for, we'll
+// calculate it's value as a linear extrapolation.
 type Distribution map[int]int64
 
-/**
- * Ensure that keys and values are in sorted order.
- */
+// CheckValidity ensures that keys and values are in sorted order.
 func (dist *Distribution) CheckValidity() error {
 	lastSeenKey := 0
 	lastSeenValue := int64(0)
@@ -34,19 +30,17 @@ func (dist *Distribution) CheckValidity() error {
 			lastSeenKey = key
 			lastSeenValue = (*dist)[key]
 		} else {
-			return fmt.Errorf("%d >= %d and %d >= %d did not hold for %s", key, lastSeenKey, (*dist)[key], lastSeenValue, *dist)
+			return fmt.Errorf("%d >= %d and %d >= %d did not hold for %v", key, lastSeenKey, (*dist)[key], lastSeenValue, *dist)
 		}
 	}
 
 	return nil
 }
 
-/**
- * Ensures that a key and value will fit properly into a distribution
- * that doesn't already contain them.
- * If the distribution already contains the key/value pair, then
- * return immediately.
- */
+// CheckKeyValueFits ensures that a key and value will fit properly
+// into a distribution that doesn't already contain them. If the
+// distribution already contains the key/value pair, then
+// return immediately.
 func (dist *Distribution) CheckKeyValueFits(key int, value int64) error {
 	if (*dist)[key] == value {
 		return nil
@@ -73,12 +67,10 @@ func (dist *Distribution) CheckKeyValueFits(key int, value int64) error {
 	return nil
 }
 
-/**
- * Returns the keys in the distribution sorted numerically.
- */
+// SortedKeys returns the keys in the distribution sorted numerically.
 func (dist *Distribution) SortedKeys() []int {
 	var keys = []int{}
-	for key, _ := range *dist {
+	for key := range *dist {
 		keys = append(keys, key)
 	}
 
@@ -87,13 +79,11 @@ func (dist *Distribution) SortedKeys() []int {
 	return keys
 }
 
-/**
- * Returns a key pair that is directly above and below the requested
- * key.
- *
- * Precondition: the Distribution must have hadd AddMinMax() called
- * and pass a CheckValidity() test
- */
+// FindHighLowKeys returns a key pair that is directly above and below the requested
+// key.
+//
+// Precondition: the Distribution must have hadd AddMinMax() called
+// and pass a CheckValidity() test
 func (dist *Distribution) FindHighLowKeys(key int) (int, int) {
 	low := 0
 	high := 0
@@ -114,31 +104,25 @@ func (dist *Distribution) FindHighLowKeys(key int) (int, int) {
 	return high, low
 }
 
-/**
- * If the key doesn't exist, returns a linear extrapolation based on
- * its position.
- */
+// Get returns a linear extrapolation based of the value if the key exists.
 func (dist *Distribution) Get(untrustedKey int) int64 {
 	// Ensures that the key is in the range [0,1000]
 	requested := int(math.Min(math.Max(float64(untrustedKey), 0), 1000))
 	if value, ok := (*dist)[requested]; ok {
 		return value
-	} else {
-		high, low := dist.FindHighLowKeys(requested)
-		highValue := (*dist)[high]
-		lowValue := (*dist)[low]
-
-		// Reminder: percentiles are multiplied by 10, values are not.
-		increment := float64(highValue-lowValue) / (float64(high-low) / float64(10))
-		delta := float64((requested-low)/10) * increment
-
-		return lowValue + int64(delta)
 	}
+	high, low := dist.FindHighLowKeys(requested)
+	highValue := (*dist)[high]
+	lowValue := (*dist)[low]
+
+	// Reminder: percentiles are multiplied by 10, values are not.
+	increment := float64(highValue-lowValue) / (float64(high-low) / float64(10))
+	delta := float64((requested-low)/10) * increment
+
+	return lowValue + int64(delta)
 }
 
-/**
- * Ensures that the Distribution's min/max are set properly.
- */
+// AddMinMax ensures that the Distribution's min/max are set properly.
 func (dist *Distribution) AddMinMax() {
 	// Always set min to 0.
 	(*dist)[0] = 0
@@ -156,14 +140,12 @@ func (dist *Distribution) AddMinMax() {
 	}
 }
 
-/**
- * Given a map of three-digit percentiles to int64 values, returns
- * a validated Distribution.
- *
- * This function expects the percentiles to be mapped from the
- * two-digit space into the three-digit space. so instead of 99.9
- * you'd pass in 999.
- */
+// FromMap takes a map of three-digit percentiles to int64 values and returns
+// a validated Distribution.
+//
+// This function expects the percentiles to be mapped from the
+// two-digit space into the three-digit space. so instead of 99.9
+// you'd pass in 999.
 func FromMap(m map[int]int64) (Distribution, error) {
 	dist := Empty()
 
@@ -175,11 +157,12 @@ func FromMap(m map[int]int64) (Distribution, error) {
 
 	if err := dist.CheckValidity(); err == nil {
 		return dist, nil
-	} else {
-		return nil, err
 	}
+
+	return nil, err
 }
 
+// Empty returns a blank Distribution.
 func Empty() Distribution {
 	return Distribution{}
 }

--- a/client/main.go
+++ b/client/main.go
@@ -2,13 +2,16 @@ package main
 
 import (
 	pb "../protos"
+	"encoding/json"
 	"flag"
 	"fmt"
 	"github.com/buoyantio/strest-grpc/client/distribution"
 	"github.com/buoyantio/strest-grpc/client/percentiles"
+	"github.com/codahale/hdrhistogram"
 	"golang.org/x/net/context"
 	"google.golang.org/grpc"
 	"log"
+	"math"
 	"math/rand"
 	"os"
 	"os/signal"
@@ -18,10 +21,11 @@ import (
 	"time"
 )
 
-const (
-	address     = "localhost:11111"
-	defaultName = "world"
-)
+// MeasuredResponse tracks the latency of a response and any accompaning error.
+type MeasuredResponse struct {
+	latency time.Duration
+	err     error
+}
 
 func exUsage(msg string) {
 	fmt.Fprintf(os.Stderr, "%s\n", msg)
@@ -30,11 +34,55 @@ func exUsage(msg string) {
 	os.Exit(64)
 }
 
+// Report provides top-level good/bad/failed numbers and a latency breakdown.
+type Report struct {
+	Good    int64    `json:"good"`
+	Bad     int64    `json:"bad"`
+	Failed  int64    `json:"failed"`
+	Latency *Latency `json:"latency"`
+}
+
+// Latency Percentiles
+type Latency struct {
+	Quantile50  int64 `json:"50"`
+	Quantile95  int64 `json:"95"`
+	Quantile99  int64 `json:"99"`
+	Quantile999 int64 `json:"999"`
+}
+
+func logFinalReport(good int64, bad int64, failed int64, histogram *hdrhistogram.Histogram) {
+	latency := Latency{
+		Quantile50:  histogram.ValueAtQuantile(50) / 1000000,
+		Quantile95:  histogram.ValueAtQuantile(95) / 1000000,
+		Quantile99:  histogram.ValueAtQuantile(99) / 1000000,
+		Quantile999: histogram.ValueAtQuantile(999) / 1000000,
+	}
+
+	report := Report{
+		Good:    good,
+		Bad:     bad,
+		Failed:  failed,
+		Latency: &latency,
+	}
+
+	if data, err := json.MarshalIndent(report, "", "  "); err != nil {
+		log.Fatal("Unable to generate report: ", err)
+	} else {
+		fmt.Println(string(data))
+	}
+}
+
 func main() {
-	concurrency := flag.Int("concurrency", 1, "client concurrency level")
-	requests := flag.Int("requests", 100, "number of requests per connection")
-	latencyPercentileFlag := flag.String("latencyPercentiles", "0=10,100=1000", "response latency percentile distribution.")
-	lengthPercentileFlag := flag.String("lengthPercentiles", "0=100,100=10000", "response body length percentile distribution.")
+	var (
+		address               = flag.String("address", "localhost:11111", "hostname:port of strest-grpc service or intermediary")
+		concurrency           = flag.Int("concurrency", 1, "client concurrency level")
+		requests              = flag.Int("requests", 10000, "number of requests per connection")
+		interval              = flag.Duration("interval", 10*time.Second, "reporting interval")
+		latencyPercentileFlag = flag.String("latencyPercentiles", "50=10,100=100", "response latency percentile distribution.")
+		lengthPercentileFlag  = flag.String("lengthPercentiles", "50=100,100=1000", "response body length percentile distribution.")
+		disableFinalReport    = flag.Bool("disableFinalReport", false, "do not print a final JSON output report")
+		onlyFinalReport       = flag.Bool("onlyFinalReport", false, "only print the final report, nothing intermediate")
+	)
 
 	flag.Usage = func() {
 		fmt.Fprintf(os.Stderr, "Usage: %s [flags]\n", path.Base(os.Args[0]))
@@ -42,6 +90,10 @@ func main() {
 	}
 
 	flag.Parse()
+
+	if *onlyFinalReport && *disableFinalReport {
+		log.Fatalf("cannot use both -onlyFinalReport and -disableFinalReport.")
+	}
 
 	if *concurrency < 1 {
 		exUsage("concurrency must be at least 1")
@@ -67,17 +119,29 @@ func main() {
 		log.Fatalf("unable to create length distribution: %v", err)
 	}
 
-	r := rand.New(rand.NewSource(time.Now().UnixNano()))
 	cleanup := make(chan os.Signal)
 	signal.Notify(cleanup, syscall.SIGINT)
+
+	var count, totalCount, good, totalGood, failed, totalFailed, bad, totalBad, max, min int64
+	min = math.MaxInt64
+
+	hist := hdrhistogram.New(0, int64(time.Second), 5)
+	globalHist := hdrhistogram.New(0, int64(time.Second), 5)
+	received := make(chan *MeasuredResponse, 10000)
+
+	var timeout = make(<-chan time.Time)
+	if !*onlyFinalReport {
+		timeout = time.After(*interval)
+	}
 
 	var wg sync.WaitGroup
 	wg.Add(*concurrency)
 
 	for i := int(0); i < *concurrency; i++ {
 		go func() {
+			r := rand.New(rand.NewSource(time.Now().UnixNano()))
 			// Set up a connection to the server.
-			conn, err := grpc.Dial(address, grpc.WithInsecure())
+			conn, err := grpc.Dial(*address, grpc.WithInsecure())
 			if err != nil {
 				log.Fatalf("did not connect: %v", err)
 			}
@@ -87,15 +151,18 @@ func main() {
 			for j := int(0); j < *requests; j++ {
 				lengthValue := lengthDistribution.Get(r.Int() % 1000)
 				latencyValue := latencyDistribution.Get(r.Int() % 1000)
-				r, err := c.Get(context.Background(),
+				start := time.Now()
+				_, err := c.Get(context.Background(),
 					&pb.ResponseSpec{
 						Count:   1,
 						Length:  int32(lengthValue),
 						Latency: latencyValue})
+
+				received <- &MeasuredResponse{time.Since(start), err}
+
 				if err != nil {
 					log.Fatalf("could not send a request: %v", err)
 				}
-				log.Printf("Response: %s", r.Body)
 			}
 			wg.Done()
 		}()
@@ -105,12 +172,57 @@ func main() {
 		for {
 			select {
 			case <-cleanup:
-				go func() {
-					os.Exit(1)
-				}()
+				if !*disableFinalReport {
+					logFinalReport(totalGood, totalBad, totalFailed, globalHist)
+				}
+				os.Exit(1)
+			case response := <-received:
+				count++
+				totalCount++
+				if response.err != nil {
+					failed++
+					totalFailed++
+				} else {
+					good++
+					totalGood++
+					latency := response.latency.Nanoseconds()
+					if latency < min {
+						min = latency
+					}
+
+					if latency > max {
+						max = latency
+					}
+
+					hist.RecordValue(latency)
+					globalHist.RecordValue(latency)
+				}
+			case t := <-timeout:
+				if min == math.MaxInt64 {
+					min = 0
+				}
+				// Periodically print stats about the request load.
+				fmt.Printf("%s %6d/%1d/%1d requests %s %3d [%3d %3d %3d %4d ] %4d\n",
+					t.Format(time.RFC3339),
+					good,
+					bad,
+					failed,
+					interval,
+					min/1000000,
+					hist.ValueAtQuantile(50)/1000000,
+					hist.ValueAtQuantile(95)/1000000,
+					hist.ValueAtQuantile(99)/1000000,
+					hist.ValueAtQuantile(999)/1000000,
+					max/1000000)
+				count, good, bad, failed, max, min = 0, 0, 0, 0, 0, math.MaxInt64
+				hist.Reset()
+				timeout = time.After(*interval)
 			}
 		}
 	}()
 
 	wg.Wait()
+	if !*disableFinalReport {
+		logFinalReport(totalGood, totalBad, totalFailed, globalHist)
+	}
 }

--- a/client/percentiles/percentiles.go
+++ b/client/percentiles/percentiles.go
@@ -5,31 +5,27 @@ import (
 	"strings"
 )
 
-/**
- * Given a two-digit percentile value, converts it to a three-digit
- * value
- */
+// ConvertToBetterPercentile takes a two-digit percentile value
+// and converts it to a three-digit value
 func ConvertToBetterPercentile(p int) int {
 	if p > 100 {
 		return p
-	} else {
-		return p * 10
 	}
+
+	return p * 10
 }
 
-/**
- * Given an input of key-value pairs corresponding to percentiles and
- * values, return a map of them. Converts percentiles from two digits
- * to three digits to account for percentiles that aren't whole
- * numbers e.g. 99.9
- *
- * Ex: 50=100,90=200,95=1000,999=20000, we
- * return the following map:
- * p[500] = 100
- * p[900] = 200
- * p[950] = 1000
- * p[999] = 20000
- */
+// ParsePercentiles takes a string of key-value pairs corresponding to
+// percentiles and  values, return a map of them. Converts percentiles
+// from two digits to three digits to account for percentiles that
+// aren't whole numbers e.g. 99.9
+//
+// Ex: 50=100,90=200,95=1000,999=20000, we
+// return the following map:
+// p[500] = 100
+// p[900] = 200
+// p[950] = 1000
+// p[999] = 20000
 func ParsePercentiles(input string) (map[int]int64, error) {
 	var percentiles map[int]int64
 	var ss []string

--- a/server/random_string/random_string.go
+++ b/server/random_string/random_string.go
@@ -4,12 +4,6 @@ import (
 	"math/rand"
 )
 
-/**
- * Fast Random String generation borrowed from
- *  http://stackoverflow.com/a/31832326/793696
- *  https://play.golang.org/p/WIgH7GRnN1
- */
-
 const letterBytes = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
 const (
 	letterIdxBits = 6                    // 6 bits to represent a letter index
@@ -17,6 +11,9 @@ const (
 	letterIdxMax  = 63 / letterIdxBits   // # of letter indices fitting in 63 bits
 )
 
+// Fast Random String generation borrowed from
+//  http://stackoverflow.com/a/31832326/793696
+//  https://play.golang.org/p/WIgH7GRnN1
 func RandStringBytesMaskImprSrc(src rand.Source, n int) string {
 	b := make([]byte, n)
 	// A src.Int63() generates 63 random bits, enough for letterIdxMax characters!

